### ISSUE TITLE
TestActivePages.testBasic fails on Python bridge

### ISSUE THREAD

**Issue by jelly** [cockpit-project/cockpit#18244]
### Explain what happens

The test disables preloading of the terminal page, loads the terminal page, spawns `basc -c 'exec -a kitten cat'` and then disables the active terminal page. The terminal iframe goes away, but the bridge does not kill the `kitten` process.

https://cockpit-logs.us-east-1.linodeobjects.com/pull-18239-20230125-083800-1422b606-fedora-37-pybridge/log.html#323

**Comment by jelly** [comment]
@martinpitt what would normally send an close signal to cockpit-ws and then to cockpit-bridge to close running processes?

**Comment by martinpitt** [comment]
I remember looking at this two weeks ago. I was hoping that commit e100e4d8b0b4bd53f28fa746003f7439246d4662 would fix it, but it didn't. Aside from that I'm afraid I don't have any idea right now how these dangling subprocesses should be killed.. I can have a closer look today.

**Comment by martinpitt** [comment]
With the C bridge, the process inheritance looks like this:
```
systemd(1)-cockpit-ws(4757)-+-cockpit-session(4766)---cockpit-bridge(4783)-+-bash(6598)---cat(6626)
```
the last is our "kitten", but pstree doesn't recognize that. I.e. this isn't spawned directly by the bridge, but by the bash running inside the Terminal. When killing the terminal page, this happens to kitten:
```
6626  read(0, 0x7f99c6b88000, 131072)   = -1 EIO (Input/output error)
6626  --- SIGHUP {si_signo=SIGHUP, si_code=SI_USER, si_pid=6598, si_uid=1000} ---
6626  +++ killed by SIGHUP +++
```
and this to the parent bash:
```
6598  wait4(-1, 0x7ffce3f948e0, WSTOPPED|WCONTINUED, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
6598  --- SIGHUP {si_signo=SIGHUP, si_code=SI_KERNEL} ---
6598  --- SIGCONT {si_signo=SIGCONT, si_code=SI_KERNEL} ---
6598  rt_sigreturn({mask=[CHLD]})       = -1 EINTR (Interrupted system call)
[... some irrelevant /home/admin/.bash_history updating removed ...]
6598  kill(-6626, SIGHUP)               = 0
6598  rt_sigprocmask(SIG_BLOCK, [CHLD TSTP TTIN TTOU], [CHLD], 8) = 0
6598  ioctl(255, TIOCSPGRP, [6598])     = -1 ENOTTY (Inappropriate ioctl for device)
6598  rt_sigprocmask(SIG_SETMASK, [CHLD], NULL, 8) = 0
6598  setpgid(0, 6598)                  = -1 EPERM (Operation not permitted)
6598  rt_sigprocmask(SIG_BLOCK, [CHLD], [CHLD], 8) = 0
6598  rt_sigprocmask(SIG_SETMASK, [CHLD], NULL, 8) = 0
6598  rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
6598  rt_sigaction(SIGHUP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7fa67943ea30}, {sa_handler=0x55c87a8d6ef0, sa_mask=[HUP INT ILL TRAP ABRT BUS FPE USR1 SEGV USR2 PIPE ALRM TERM XCPU XFSZ VTALRM SYS], sa_flags=SA_RESTORER, sa_restorer=0x7fa67943ea30}, 8) = 0
6598  getpid()                          = 6598
6598  --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=6263, si_uid=1000} ---
6598  kill(6598, SIGHUP)                = 0
6598  --- SIGHUP {si_signo=SIGHUP, si_code=SI_USER, si_pid=6598, si_uid=1000} ---
6598  +++ killed by SIGHUP +++
```
In plain English, the terminal shell receives a SIGHUP, which almost certainly is the `prctl (PR_SET_PDEATHSIG, SIGHUP)` that we do in cockpitpipe.c. It then kills its children with SIGHUP as well, and finally itself.

OTOH, with the pybridge, the Terminal shell does absolutely nothing. strace just shows the pending `wait4(-1,  <detached ...>` call (waiting for the executing `kitten` to finish), and kitten just does `read(0,  <detached ...>`, i.e. waits on its stdin for anything to happen. Thus, killing the terminal page does not SIGHUP its children.

I have an strace of the C bridge, but digging through that will take some time.

**Comment by martinpitt** [comment]
So the problem isn't around PDEATHSIG -- the "bash" process from the terminal isn't killed itself, so there isn't any PDEATHSIG to send to the kitten child. When the shell kills the frame, it sends

    send control: {"group":"cockpit1:localhost/system/terminal","command":"kill"}

The py bridge does implement the kill command and `group`, but not sufficiently? Let's put that into a unit test.

**Comment by martinpitt** [comment]
The above approach with using `cockpit.transport.control()` is broken, so I am now reproducing this locally with

    ./cockpit-ws -p 9999 --local-session=./b.sh

with `b.sh`:
```sh
COCKPIT_DEBUG=all PYTHONPATH=src python3 -m cockpit.bridge
```

and this additional patch to more forcefully close the channel:
```diff
--- src/cockpit/router.py
+++ src/cockpit/router.py
@@ -153,8 +153,11 @@ class Router(CockpitProtocolServer):
 
     def do_kill(self, host: Optional[str], group: Optional[str]) -> None:
         if group:
+            logger.debug('Router: Killing group %s', group)
             to_close = set(ch for ch, gr in self.groups.items() if gr == group)
             for channel in to_close:
+                logger.debug('  killing channel %s', channel)
+                self.open_channels[channel].close(problem='terminated')
                 self.close_channel(channel)
 
     def channel_control_received(self, channel: str, command: str, message: Dict[str, object]) -> None:

```

Go to http://localhost:9999/system/terminal , use Alt+click menu, and kill the terminal frame.

```
cockpit.protocol-DEBUG: channel control received {'payload': 'stream', 'spawn': ['/bin/bash'], 'environ': ['TERM=xterm-256color'], 'directory': '/home/martin', 'pty': True, 'command': 'open', 'channel': '1:1!2', 'host': 'localhost', 'flow-control': True, 'group': 'cockpit1:localhost/system/terminal'}
[... kill the frame ... ]
cockpit.protocol-DEBUG: transport control received {'group': 'cockpit1:localhost/system/terminal', 'command': 'kill'}
cockpit.router-DEBUG: Router: Killing group cockpit1:localhost/system/terminal
cockpit.router-DEBUG:   killing channel 1:1!1
cockpit.protocol-DEBUG: sending message  {'channel': '1:1!1', 'command': 'close', 'problem': 'terminated'}
cockpit.protocol-DEBUG: writing to transport <cockpit.transports.StdioTransport object at 0x7fb17dd0f790>
cockpit.router-DEBUG:   killing channel 1:1!2
cockpit.protocol-DEBUG: sending message  {'channel': '1:1!2', 'command': 'close', 'problem': 'terminated'}
cockpit.protocol-DEBUG: writing to transport <cockpit.transports.StdioTransport object at 0x7fb17dd0f790>
```
So this confirms that it *tries* to close the correct channel for terminal bash (`1:1!2`), but even with the extra patch it does not *actually* do that.
